src/name_utils.rs

```diff
@@ -0,0 +1,253 @@
+            SubName::Group => GROUP_NAME_MAX_LEN,
+            SubName::Node => NODE_NAME_MAX_LEN,
+            SubName::Misc => MISC_MAX_LEN,
+        }
+    }
+}
+
+/// Build a Kubernetes resource name. This is intended to work with the generatedName (add a unique
+/// hash) from Kubernetes. This method ensures that all single components do not exceed a certain
+/// length in order to keep the resource name below 63 (the maximum allowed characters) minus an
+/// offset for the unique hash.
+///
+/// In each sub name (`short_name`, `cluster_name`...) all non alphanumeric parts are automatically
+/// removed. After the removal each sub_name is cut off if it exceeds its specified length.
+///
+/// After processing each sub name is concatenated with "-" and "-" as the last character (to
+/// separate from the kubernetes hash).
+///
+/// # Arguments
+///
+/// * `short_name` - The short name of the custom resource definition.
+/// * `cluster_name` - The name of the custom resource.
+/// * `role_name` - The role name of the custom resource.
+/// * `group_name` - The group name of the custom resource.
+/// * `node_name` - Optional node name if available (e.g. for pods).
+/// * `misc_name` - Optional miscellaneous identifiers (e.g. "data" for config maps).
+///
+pub fn build_resource_name(
+    short_name: &str,
+    cluster_name: &str,
+    role_name: &str,
+    group_name: &str,
+    node_name: Option<&str>,
+    misc_name: Option<&str>,
+) -> OperatorResult<String> {
+    let mut full_name = String::new();
+
+    full_name.push_str(&strip(SubName::Short, short_name)?);
+    full_name.push('-');
+
+    full_name.push_str(&strip(SubName::Cluster, cluster_name)?);
+    full_name.push('-');
+
+    full_name.push_str(&strip(SubName::Role, role_name)?);
+    full_name.push('-');
+
+    full_name.push_str(&strip(SubName::Group, group_name)?);
+    full_name.push('-');
+
+    if let Some(node) = node_name {
+        full_name.push_str(&strip(SubName::Node, node)?);
+        full_name.push('-');
+    };
+
+    if let Some(misc) = misc_name {
+        full_name.push_str(&strip(SubName::Misc, misc)?);
+        full_name.push('-');
+    };
+
+    Ok(full_name.to_lowercase())
+}
+
+/// This method removes all non alphanumeric characters from a sub name, checks if the
+/// length after the removal is not below the minimum (throws error) or over the maximum
+/// (will be cut off) specified length.
+///
+/// # Arguments
+///
+/// * `kind` - The kind of the sub_name (e.g. ShortName, ClusterName...).
+/// * `sub_name` - The sub_name to be processed.
+///
+fn strip(kind: SubName, sub_name: &str) -> OperatorResult<String> {
+    let alphanumeric: String = sub_name.chars().filter(|c| c.is_alphanumeric()).collect();
+
+    let min = kind.min();
+    if alphanumeric.len() < usize::from(min) {
+        return Err(Error::SubNameTooShort {
+            kind: kind.to_string(),
+            name: sub_name.to_string(),
+            min,
+        });
```

Comment:
How about padding the sub_name with nulls (0) for example instead of erroring because it's too small ? It's what the caller would probably have to do anyway.